src/Microsoft.TestPlatform.CommunicationUtilities/JsonDataSerializer.cs

```diff
@@ -81,7 +135,106 @@ public T DeserializePayload<T>(Message message)
+                // This is still better than deserializing the JToken in DeserializeMessage because here we know that the payload
+                // will actually be used.
+                return Deserialize<T>(payloadSerializer, Deserialize<VersionedMessage>(rawMessage).Payload);
+            }
+        }
+        else
+        {
+            return Deserialize<T>(payloadSerializer, message.Payload);
+        }
+    }
+
+    private bool FastHeaderParse(string rawMessage, out int version, out string messageType)
+    {
+        // PERF: This can be also done slightly better using ReadOnlySpan<char> but we don't have that available by default in .NET Framework
+        // and the speed improvement does not warrant additional dependency. This is already taking just few ms for 10k messages.
+        version = 0;
+        messageType = null;
+
+        try
+        {
+            // {"Version":6,"MessageType":"TestExecution.GetTestRunnerProcessStartInfoForRunAll","Payload":{
+            if (rawMessage[2] != 'V')
+            {
+                return false;
+            }
+
+            // Version ':' is on index 10, the number starts at the next index. Find wher the next ',' is and grab that as number.
+            var versionColonIndex = 10;
+            if (rawMessage[versionColonIndex] != ':')
+            {
+                return false;
+            }
+
+            var firstVersionNumberIndex = 11;
+            if (!TryGetSubstringUntilDelimiter(rawMessage, firstVersionNumberIndex, ',', maxSearchLength: 4, out string versionString, out int delimiterIndex))
+            {
+                return false;
+            }
+
+            if (!int.TryParse(versionString, out var versionInt))
+            {
+                return false;
+            }
+
+            // Message type delmiter is at at versionCommaIndex + the length of ',"MessageType":"' which is 16 chars
+            var messageTypeQuoteIndex = delimiterIndex + 16;
+
+            // "TestExecution.LaunchAdapterProcessWithDebuggerAttachedCallback" is the longest message we currently have with 62 chars
+            if (!TryGetSubstringUntilDelimiter(rawMessage, messageTypeQuoteIndex, '"', maxSearchLength: 100, out string messageTypeString, out _))
+            {
+                return false;
+            }
+
+
+            version = versionInt;
+            messageType = messageTypeString;
+            return true;
+        }
+        catch
+        {
+            return false;
+        }
+    }
+
+    /// <summary>
+    ///  Try getting substring until a given delimiter, but don't search more characters than maxSearchLength.
+    /// </summary>
+    private bool TryGetSubstringUntilDelimiter(string rawMessage, int start, char character, int maxSearchLength, out string substring, out int delimiterIndex)
+    {
+        var length = rawMessage.Length;
+        var searchEnd = start + maxSearchLength;
+        for (int i = start; i < length && i <= searchEnd; i++)
```

Comment:
Right, I was looking exactly for this but did not find it, and I was surprised that it is not there :D Not sure if it would be any faster, and if I get the count wrong it will throw. v2 maybe? :)

Reply:
> Not sure if it would be any faster

It's faster because it's vectorized https://github.com/dotnet/runtime/blob/181bf3122278f9b6ae1cd755e2ab99dfe1706604/src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Char.cs#L518-L794 but I think that it's ok as-is here we don't have numbers to justify atm.